src/Lib.P2P/Transports/Tcp.cs

```diff
@@ -0,0 +1,293 @@
+               .FirstOrDefault();
+            
+            var ip = address.Protocols
+               .FirstOrDefault(p => p.Name == "ip4" || p.Name == "ip6");
+            
+            if (ip == null)
+            {
+                throw new ArgumentException($"Missing IP address in '{address}'.", nameof(address));
+            }
+            
+            var ipAddress = IPAddress.Parse(ip.Value);
+            var endPoint = new IPEndPoint(ipAddress, port);
+            var socket = new Socket(
+                endPoint.AddressFamily,
+                SocketType.Stream,
+                ProtocolType.Tcp);
+            try
+            {
+                socket.Bind(endPoint);
+                socket.Listen(100);
+            }
+            catch (Exception e)
+            {
+                socket.Dispose();
+                throw new Exception("Bind/listen failed on " + address, e);
+            }
+
+            // If no port specified, then add it.
+            var actualPort = ((IPEndPoint) socket.LocalEndPoint).Port;
+            
+            if (port != actualPort)
+            {
+                address = address.Clone();
+                var protocol = address.Protocols.FirstOrDefault(p => p.Name == "tcp");
+                if (protocol != null)
+                {
+                    protocol.Value = actualPort.ToString();
+                }
+                else
+                {
+                    address.Protocols.AddRange(new MultiAddress("/tcp/" + actualPort).Protocols);
+                }
+            }
+
+            _ = Task.Run(() => ProcessConnection(socket, address, handler, cancel), cancel);
+
+            return address;
+        }
+
+        private static void ProcessConnection(Socket socket,
+            MultiAddress address,
+            Action<Stream, MultiAddress, MultiAddress> handler,
+            CancellationToken cancel)
+        {
+            Log.Debug("listening on " + address);
+
+            // Handle cancellation of the listener
+            cancel.Register(() =>
+            {
+                Log.Debug("Got cancel on " + address);
+
+                try
+                {
+                    if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+                    {
+                        socket.Shutdown(SocketShutdown.Both);
+                        socket.Dispose();
+                    }
+                    else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+                    {
+                        socket.Shutdown(SocketShutdown.Receive);
+                    }
+                    else // must be windows
+                    {
+                        socket.Dispose();
+                    }
+                }
+                catch (Exception e)
+                {
+                    Log.Warn($"Cancelling listener: {e.Message}");
+                }
+                finally
+                {
+                    socket = null;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Introduce a new variable instead of reusing the parameter 'socket'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)